### ISSUE TITLE
Sub-domains treated as third-party?

### ISSUE THREAD

**Issue by Graham-sr** [craigmiskell/cookiemaster#22]
I have been using this extension for about a month and find that it works well. I have also followed your excellent explanation of how it operates. However I am confused over the handling of some sub-domains that seem to be regarded as "third-party", but NOT identified as being in an iframe.
This happens on several sites, but my examples are for American Express. Note, I log in using the UK site - https://www.americanexpress.com/en-gb/account/login - the cookies may be different for other countries.
The first 2 images show the situation for the default "Block All by Policy", before and after the main cookie is set as approved.
I then set the third-party rule to "Allow only if explicitly permitted" and the 2 "Blocked" images show the situation before and after the main cookie is set as approved.
So, my query is why are these sub-domains being regarded as third-party?
Additionally I am confused regarding the .americanexpress.com cookie. I have seen similar occurrences for other sites – cookies showing both with and without the leading period ( . ). What does this mean? Is this just another sub-domain?

![Block All by Policy1](https://user-images.githubusercontent.com/65815519/105688087-15c08580-5ef1-11eb-99a4-df9c2e6a591c.png)
![Block All by Policy2](https://user-images.githubusercontent.com/65815519/105688127-25d86500-5ef1-11eb-8081-95ffeea42ff1.png)
![Blocked1](https://user-images.githubusercontent.com/65815519/105688217-3d175280-5ef1-11eb-84fc-c856d30f8ef8.png)
![Blocked2](https://user-images.githubusercontent.com/65815519/105688228-41437000-5ef1-11eb-802c-cd8eebb6ae3c.png)

**Comment by craigmiskell** [comment]
Sorry for the delay.  

In short, because cookiemaster currently thinks a cookie is third party if it is set on an HTTP request to a resource (e.g image, css, etc) where the hostname of the resource isn't *exactly* the same as the tab URL.  So on www.americanexpress.com, any cookies set during a request to global.americanexpress.com are 3rd party.  This is implemented at https://github.com/craigmiskell/cookiemaster/blob/master/background.js#L155

The .americanexpress.com cookie is being set during a request to global.americanexpress.com. I had to fix the logging to actually report this, but it ends up looking like:
```
8/02/2021, 5:30:48 pm	INFO	3898: Blocking third party cookie for domain global.americanexpress.com in header of request on www.americanexpress.com
8/02/2021, 5:30:48 pm	INFO	3898: Blocking third party cookie for domain .americanexpress.com in header of request on www.americanexpress.com
8/02/2021, 5:30:48 pm	INFO	3898: Blocking third party cookie for domain global.americanexpress.com in header of request on www.americanexpress.com
8/02/2021, 5:30:48 pm	DEBUG	3898: set-cookie for https://global.americanexpress.com/api/servicing/v1/member on tab with url https://www.americanexpress.com/en-gb/account/login
```

Logs are in reverse order, so a request was made to `https://global.americanexpress.com/api/servicing/v1/member`, where the tab was `https://www.americanexpress.com/en-gb/account/login`, so cookiemaster said all 3 cookies that it tried to set (including one on .americanexpress.com) were 3rd party, and thus blocked.

Indeed, the (quite old) comment immediately after line 155  calls out specifically this sort of case as worth more thinking, which I've not done yet.  After some hunting, https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/mozIThirdPartyUtil#isThirdPartyURI() is the best looking definition of 3rd party I can find:  
> Determine whether two URIs are third party with respect to each other. This is determined by computing the base domain for both URIs. If they can be determined, and the base domains match, the request is defined as first party. If it cannot be determined because one or both URIs do not have a base domain (for instance, in the case of IP addresses, host aliases such as 'localhost', or a file:// URI), an exact string comparison on host is performed.


From the [implementation](https://searchfox.org/comm-central/source/mozilla/dom/base/ThirdPartyUtil.cpp#398-406), it appears Mozilla is using eTLDs (the [public suffix](https://publicsuffix.org/) list) as the core of that.    Thankfully cookiemaster already has that, but is only using it for display purposes.  It shouldn't be too hard to rejig the definition of what is third party to better match Firefox.  Might be a day or two, but I will definitely fix it up.

**Comment by craigmiskell** [comment]
Just released 1.2.0.  It should be available shortly.
